### PR TITLE
[bug] Replace_sampler attach previous multiprocessing_context

### LINKED ISSUE #4486: TrainerDataLoadingMixin.replace_sampler ignores multiprocessing_context
## 🐛 Bug

replace_sampler in TrainerDataLoadingMixin ignores multiprocessing_context

## Please reproduce using [the BoringModel and post here]

https://colab.research.google.com/drive/1IRCGh1HbutELn3ADsZ9jAwDjc6zoyB_R?usp=sharing

### Expected behavior

It should return a new data loader with the replaced sampler, and the same multiprocessing_context

### Environment

* CUDA:
	- GPU:
		- Tesla T4
	- available:         True
	- version:           10.1
* Packages:
	- numpy:             1.18.5
	- pyTorch_debug:     False
	- pyTorch_version:   1.6.0+cu101
	- pytorch-lightning: 0.10.0
	- tqdm:              4.41.1
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- 
	- processor:         x86_64
	- python:            3.6.9
	- version:           #1 SMP Thu Jul 23 08:00:38 PDT 2020

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Fixes #4486

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In in short, see following bullet-list:  

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by SeanNaren — APPROVED** [review]

File: tests/trainer/test_dataloaders.py [hunk 1]
```diff
@@ -1112,3 +1114,27 @@ def test_dataloaders_load_only_once_passed_loaders(tmpdir):
...
     for call, expected in zip(calls, expected_sequence):
         assert call['name'] == expected
+
+
+def test_replace_sampler_with_multiprocessing_context(tmpdir):
+    """
+    This test verifies that replace_sampler conserves multiprocessing context
+    """
+    train = RandomDataset(32, 64)
+    train = DataLoader(train, batch_size=32, num_workers=2, multiprocessing_context='spawn', shuffle=True)
```
Comment (Borda) [inline tests/trainer/test_dataloaders.py:1124]:
```suggestion
    context = 'spawn'
    train = DataLoader(train, batch_size=32, num_workers=2, multiprocessing_context=context, shuffle=True)
```

File: tests/trainer/test_dataloaders.py [hunk 2]
```diff
@@ -1112,3 +1114,27 @@ def test_dataloaders_load_only_once_passed_loaders(tmpdir):
...
+    trainer = Trainer(
+        max_epochs=1,
+        progress_bar_refresh_rate=20,
+        overfit_batches=5,
+    )
+
+    new_data_loader = trainer.replace_sampler(train, SequentialSampler(train.dataset))
+    # Train the model ⚡
+    # print(new_data_loader)
+    assert (new_data_loader.multiprocessing_context == train.multiprocessing_context)
```
Comment (Borda) [inline tests/trainer/test_dataloaders.py:1140]:
```suggestion
    assert new_data_loader.multiprocessing_context == context
```

**Review by Borda — COMMENTED** [review]
does the `multiprocessing_context` copy content or just pointer?

File: tests/trainer/test_dataloaders.py (hunk 2 above)
Comment (tchaton) [reply tests/trainer/test_dataloaders.py:1141]: Reference to the SpawnContext Object.
```
<multiprocessing.context.SpawnContext object at 0x12636aa50>
```

**Comment by tchaton** [comment]
Hey @williamFalcon, can you review this one ?

**Review by justusschock — APPROVED** [review]

**Review by williamFalcon — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]
